### PR TITLE
Added Top level Object Access.

### PR DESCRIPTION
**Pull request by waheedahmed**
[ECOM-4978](https://openedx.atlassian.net/browse/ECOM-4978)

As a user, I need access to all courses, course runs, people and organizations within my Group and should not be able to see objects owned by Groups that I'm not part of.

**Acceptance criteria:**
1. Validate staff user can have access to all Objects.
2. Validate that users have access to Objects within their parent Group.
3. Validate a user does NOT have access to objects outside of his/her Group.

### CODE REVIEW

**Comment by waheedahmed** [comment]
@awais786 @bderusha Please review.

File: course_discovery/apps/publisher/mixins.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
+from django.http import HttpResponseForbidden
+
+from course_discovery.apps.publisher.models import Course, CourseRun, Seat
+
+
+class ViewPermissionMixin(object):
+
+    def get_course(self):
+        course_object = self.get_object()
```
Comment (bderusha) [inline course_discovery/apps/publisher/mixins.py]: NIT: rename this variable.  It looks like we are checking to see if a course is actually a seat.  This is confusing.  Please use something more generic

File: course_discovery/apps/publisher/mixins.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        if isinstance(course_object, CourseRun):
+            return course_object.course
+        if isinstance(course_object, Seat):
+            return course_object.course_run.course
+
+        return course_object
```
Comment (bderusha) [inline course_discovery/apps/publisher/mixins.py]: I'd rather be more explicit here.  If I use this mixin with an organization endpoint `get_course` will return the organization.  Let's do some type checking for all expected returns and return null otherwise.

Comment (bderusha) [inline course_discovery/apps/publisher/mixins.py]: A good catchall might be to try to call `.course` on whatever object is passed in if it isn't an expected type.

File: course_discovery/apps/publisher/tests/test_views.py [hunk 3]
```diff
@@ -46,8 +47,8 @@ def test_create_course(self):
         self.assertNotContains(response, 'Add new comment')
         self.assertNotContains(response, 'Total Comments')
 
-    def test_update_course(self):
-        """ Verify that we can update an existing course. """
+    def test_update_course_with_staff(self):
+        """ Verify that staff user can update an existing course. """
```
Comment (bderusha) [inline course_discovery/apps/publisher/tests/test_views.py:54]: Wondering about this change despite no changes to the test.  Is it true that only a staff user can hit this page?  Are we already accessing this page with a staff user?

File: course_discovery/apps/publisher/tests/test_views.py [hunk 4]
```diff
@@ -143,6 +219,81 @@ def test_update_course_run(self):
         self.assertContains(response, 'Add new comment')
         self.assertContains(response, comment.comment)
 
+    def test_edit_course_run_page_with_non_staff(self):
+        """ Verify that non staff user can't access course run edit page without permission. """
+        non_staff_user = UserFactory()
+        group = factories.GroupFactory()
+
+        self.client.logout()
+        self.client.login(username=non_staff_user.username, password=USER_PASSWORD)
```
Comment (bderusha) [inline course_discovery/apps/publisher/tests/test_views.py]: May be worth pulling this and the 5 lines above into its own method given the number of times it is used to setup for a non-staff user

File: course_discovery/apps/publisher/views.py [hunk 5]
```diff
@@ -25,10 +27,11 @@ class CourseRunListView(ListView):
     def get_queryset(self):
         return [
             CourseRunWrapper(course_run) for course_run in CourseRun.objects.select_related('course').all()
+            if check_view_permission(self.request.user, course_run.course)
```
Comment (bderusha) [inline course_discovery/apps/publisher/views.py]: is this a database query each time?  What is the performance of running this check against every course   run record in the db?  Can this be attached to the query at all?

**Comment by bderusha** [comment]
A couple questions, but overall this looks awesome. Nice work!

File: course_discovery/apps/publisher/tests/test_views.py (hunk 3 above)
Comment (waheedahmed) [reply course_discovery/apps/publisher/tests/test_views.py:54]: @bderusha In setup of this test case staff user is logged in, so staff user can access any course according to 1st AC of this ticket.

**Comment by waheedahmed** [comment]
@bderusha does it looks good now?

File: course_discovery/apps/publisher/mixins.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    def get_course(self):
+        publisher_object = self.get_object()
+        if isinstance(publisher_object, Course):
+            return publisher_object
+        if isinstance(publisher_object, CourseRun):
+            return publisher_object.course
+        if isinstance(publisher_object, Seat):
+            return publisher_object.course_run.course
+
+        return None
```
Comment (bderusha) [inline course_discovery/apps/publisher/mixins.py:17]: Please add a test for `None`

File: course_discovery/apps/publisher/mixins.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+from course_discovery.apps.publisher.models import Course, CourseRun, Seat
+
+
+class ViewPermissionMixin(object):
+
+    def get_course(self):
+        publisher_object = self.get_object()
+        if isinstance(publisher_object, Course):
+            return publisher_object
+        if isinstance(publisher_object, CourseRun):
```
Comment (bderusha) [inline course_discovery/apps/publisher/mixins.py]: Instead of checking for type here.  Let's make this the last check and change it to 

```
if hasattr(publisher_object, 'course'):
    return publisher_object.course
```

This is more robust and will do the right thing for future objects if that method is available without needing to add checks to this each time

File: course_discovery/apps/publisher/views.py [hunk 8]
```diff
@@ -23,12 +26,16 @@ class CourseRunListView(ListView):
     template_name = 'publisher/course_runs_list.html'
 
     def get_queryset(self):
-        return [
-            CourseRunWrapper(course_run) for course_run in CourseRun.objects.select_related('course').all()
-        ]
+        if self.request.user.is_staff:
+            course_runs = CourseRun.objects.select_related('course').all()
+        else:
+            courses = get_objects_for_user(self.request.user, Course.VIEW_PERMISSION, Course)
```
Comment (bderusha) [inline course_discovery/apps/publisher/views.py:32]: Does this return the actual objects or a lazy eval'd queryset?  If it returns something we can chain on it is probably worth doing so.  Otherwise what's here is fine.  Much improved!

File: course_discovery/apps/publisher/views.py [hunk 9]
```diff
@@ -164,11 +174,16 @@ def post(self, request, course_run_id):
         state = request.POST.get('state')
         try:
             course_run = CourseRun.objects.get(id=course_run_id)
+
+            if not check_view_permission(request.user, course_run.course):
+                return HttpResponseForbidden()
+
             course_run.change_state(target=state)
+            # pylint: disable=no-member
             messages.success(
```
Comment (bderusha) [inline course_discovery/apps/publisher/views.py:183]: Are these messages user-facing?

**Comment by bderusha** [comment]
A few more minor comments.  👍 once addressed

File: course_discovery/apps/publisher/views.py (hunk 9 above)
Comment (waheedahmed) [reply course_discovery/apps/publisher/views.py:183]: @bderusha Yes.

File: course_discovery/apps/publisher/views.py (hunk 8 above)
Comment (waheedahmed) [reply course_discovery/apps/publisher/views.py:32]: @bderusha I think it return queryset of actual objects, see [guardian.shortcuts:get_objects_for_user](http://django-guardian.readthedocs.io/en/stable/api/guardian.shortcuts.html#std:shortcut-get_objects_for_user) for more details.

**Comment by awais786** [comment]
👍
